src/clients/FtpClient.cc

```diff
@@ -69,6 +69,61 @@ escapeIAC(const char *buf)
     return ret;
 }
 
+bool
+parseEPSV(Parser::Tokenizer &tz, const SBuf &tuple, uint16_t &outPort)
+{
+    try {
+        // '(' or need more data
+        tz.skipRequired("opening '('", SBuf("("));
+
+        // Peek the delimiter from the underlying buffer (tuple[1])
+        if (tuple.length() < 2)
+            throw TextException("empty delimiter in EPSV tuple", Here());
+
+        const auto delim = static_cast<unsigned char>(tuple[1]);
+
+        // validate delimiter: ASCII 33..126 and not a digit
+        if (delim < 33 || delim > 126 || xisdigit(delim))
+            throw TextException("invalid EPSV delimiter (must be ASCII 33..126 and not a digit)", Here());
+
+        // same delimiter three times
+        if (!tz.skip(static_cast<char>(delim)) ||
+                !tz.skip(static_cast<char>(delim)) ||
+                !tz.skip(static_cast<char>(delim)))
+            throw TextException("EPSV tuple missing repeated delimiters", Here());
+
+        // 1..5 decimal digits (port)
+        const auto port64 = tz.udec64("EPSV port", 5);
+
+        // 4th delimiter and ')'
+        if (!tz.skip(static_cast<char>(delim)))
+            throw TextException("EPSV tuple missing fourth delimiter", Here());
+        tz.skipRequired("closing ')'", SBuf(")"));
+
+        // optional trailing whitespace
+        static CharacterSet ws("epsv-ws", " \t\r\n");
+        tz.skipAll(ws);
+
+        // no garbage after tuple
+        if (!tz.atEnd())
+            throw TextException("trailing garbage after EPSV tuple", Here());
+
+        // range check
+        if (port64 < 1 || port64 > 65535)
+            throw TextException(ToSBuf("EPSV port out of range: ", port64), Here());
+
+        outPort = static_cast<uint16_t>(port64);
+        return true;
+    } catch (const Parser::InsufficientInput &) {
+        debugs(9, 6, "too-short EPSV tuple");
+    } catch (const std::exception &ex) {
+        debugs(9, 6, "invalid EPSV tuple: " << ex.what());
+    } catch (...) {
+        debugs(9, 6, "cannot parse EPSV tuple: " << CurrentException);
```

Comment:
We cannot be sure only the parser syntax issues are generating `std::exception` (eg. versus memory allocations or such incidental activity). I believe it best to combine these two catch cases.
```suggestion
    } catch (...) {
        debugs(9, 6, "cannot parse EPSV tuple: " << CurrentException);
```